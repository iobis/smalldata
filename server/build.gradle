plugins {
  id 'java'
  id 'application'
  id "io.franzbecker.gradle-lombok" version "2.1"
  id 'com.github.johnrengelman.shadow' version '4.0.4'
  id 'org.javamodularity.moduleplugin' version '1.4.0'
  id 'org.beryx.runtime' version '1.1.5'
  id 'checkstyle'
  id 'pmd'
}

repositories {
  mavenLocal()
  jcenter()
}

ext {
  vertxVersion = '3.6.3'
  jacksonVersion = '2.9.8'
  junitJupiterEngineVersion = '5.2.0'
  moduleName = 'obis.smalldata'
}

sourceCompatibility = 10

group = 'org.obis'
version = '1.0.0-SNAPSHOT'

mainClassName = "io.vertx.core.Launcher"

def mainVerticleName = "org.obis.smalldata.Starter"
def watchForChange = 'src/**/*'
def doOnChange = './gradlew classes'

runtime {
  options = ['--strip-debug', '--compress', '2', '--no-header-files', '--no-man-pages']
}

dependencies {
  implementation "io.vertx:vertx-core:$vertxVersion"
  implementation "io.vertx:vertx-web:$vertxVersion"
  implementation "io.vertx:vertx-web-client:$vertxVersion"
  implementation "io.vertx:vertx-mongo-client:$vertxVersion"
  implementation "io.vertx:vertx-web-api-contract:$vertxVersion"

  implementation "com.fasterxml.jackson.dataformat:jackson-dataformat-xml:$jacksonVersion"
  implementation "com.fasterxml.jackson.datatype:jackson-datatype-jsr310:$jacksonVersion"

  implementation "de.flapdoodle.embed:de.flapdoodle.embed.mongo:2.2.0"
  implementation "org.tinylog:tinylog:1.3.5"
  implementation "org.tinylog:slf4j-binding:1.3.5"

  compileOnly "org.projectlombok:lombok:1.18.6"

  testImplementation "io.vertx:vertx-junit5:$vertxVersion"
  testImplementation "io.vertx:vertx-junit5:$vertxVersion"
  testImplementation "com.fasterxml.uuid:java-uuid-generator:3.1.5"
  testImplementation "commons-io:commons-io:2.6"

  testRuntime("org.junit.jupiter:junit-jupiter-engine:$junitJupiterEngineVersion")
}

test {
  useJUnitPlatform()
  testLogging {
    events 'PASSED', 'FAILED', 'SKIPPED'
  }
}

run {
  args = ['run', mainVerticleName, "--redeploy=$watchForChange", "--launcher-class=$mainClassName", "--on-redeploy=$doOnChange"]
}

shadowJar {
  manifest {
    attributes 'Main-Verticle': mainVerticleName
  }
  mergeServiceFiles {
    include 'META-INF/services/io.vertx.core.spi.VerticleFactory'
  }
}

checkstyle {
  ignoreFailures false
  maxErrors 0
  maxWarnings 0
  toolVersion "8.17"
}

pmd {
  consoleOutput true
  ruleSetConfig = resources.text.fromFile("config/pmd/ruleset.xml")
  ruleSets = []
}
