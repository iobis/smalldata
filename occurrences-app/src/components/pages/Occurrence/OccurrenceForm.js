import ActiveStepHeader from './StepHeaders/ActiveStepHeader'
import OccurrenceData from './OccurrenceData/OccurrenceData'
import ConfirmedStepHeader from './StepHeaders/ConfirmedStepHeader'
import DarwinCoreFields from './DarwinCoreFields/DarwinCoreFields'
import LocationData from './LocationData/LocationData'
import MeasurementOrFact from './MeasurementOrFact/MeasurementOrFact'
import NotConfirmedStepHeader from './StepHeaders/NotConfirmedStepHeader'
import ObservationData from './ObservationData/ObservationData'
import PropTypes from 'prop-types'
import React, { useState } from 'react'
import SelectDataset from './SelectDataset/SelectDataset'
import { format } from 'date-fns'
import { getDatasetMock } from '../../../clients/server'
import { useTranslation } from 'react-i18next'

export default function OccurrenceForm() {
  const datasets = getDatasetMock()
  const { t } = useTranslation()
  const [dataset, setSelectedDataset] = useState(datasets[0])
  const [occurrenceData, setOccurrenceData] = useState({
    basisOfRecord:    'humanObservation',
    beginDate:        Date.now(),
    endDate:          null,
    lifestage:        null,
    occurrenceStatus: 'present',
    scientificName:   '',
    sex:              null
  })
  const [locationData, setLocationData] = useState({
    decimalLongitude:      null,
    decimalLatitude:       null,
    coordinateUncertainty: null,
    minimumDepth:          null,
    maximumDepth:          null,
    verbatimCoordinates:   '',
    verbatimDepth:         ''
  })
  const [observationData, setObservationData] = useState({
    institutionCode:         '',
    collectionCode:          '',
    fieldNumber:             '',
    catalogNumber:           '',
    recordNumber:            '',
    identifiedBy:            [],
    recordedBy:              [],
    identificationQualifier: '',
    identificationRemarks:   '',
    references:              []
  })
  const [darwinCoreFields, setDarwinCoreFields] = useState([])
  const [activeStepIndex, setActiveStepIndex] = useState(0)
  const [measurementOrFact, setMeasurementOrFact] = useState([])
  const steps = [{
    dataDescription: t('occurrenceForm.selectDataset.step.dataDescription'),
    selectedData:    dataset.description,
    stepDescription: t('occurrenceForm.selectDataset.step.stepDescription'),
    stepTitle:       t('occurrenceForm.selectDataset.step.stepTitle'),
    children:        <SelectDataset
      datasets={datasets}
      onChange={setSelectedDataset}
      selectedDataset={dataset}/>
  }, {
    dataDescription: t('occurrenceForm.occurrenceData.step.dataDescription'),
    selectedData:    <OccurrenceDataSummary {...occurrenceData}/>,
    stepDescription: t('occurrenceForm.occurrenceData.step.stepDescription'),
    stepTitle:       t('occurrenceForm.occurrenceData.step.stepTitle'),
    children:        <OccurrenceData
      data={occurrenceData}
      onChange={setOccurrenceData}/>
  }, {
    dataDescription: t('occurrenceForm.locationData.step.dataDescription'),
    selectedData:    <SelectedLocation {...locationData}/>,
    stepDescription: t('occurrenceForm.locationData.step.stepDescription'),
    stepTitle:       t('occurrenceForm.locationData.step.stepTitle'),
    children:        <LocationData
      data={locationData}
      onChange={setLocationData}/>
  }, {
    dataDescription: t('occurrenceForm.observationData.step.dataDescription'),
    selectedData:    renderIdentifiedByLabel(observationData),
    stepDescription: t('occurrenceForm.observationData.step.stepDescription'),
    stepTitle:       t('occurrenceForm.observationData.step.stepTitle'),
    children:        <ObservationData
      observationData={observationData}
      onChange={setObservationData}/>
  }, {
    dataDescription: t('occurrenceForm.measurementOrFact.step.dataDescription'),
    selectedData:    <MeasurementOrFactSummary data={measurementOrFact}/>,
    stepDescription: t('occurrenceForm.measurementOrFact.step.stepDescription'),
    stepTitle:       t('occurrenceForm.measurementOrFact.step.stepTitle'),
    children:        <MeasurementOrFact
      data={measurementOrFact}
      onChange={setMeasurementOrFact}/>
  }, {
    dataDescription: '',
    selectedData:    '',
    stepDescription: t('occurrenceForm.darwinCoreFields.step.stepDescription'),
    stepTitle:       t('occurrenceForm.darwinCoreFields.step.stepTitle'),
    children:        <DarwinCoreFields
      fields={darwinCoreFields}
      onChange={setDarwinCoreFields}/>
  }]

  return (
    <section className="section">
      {steps.map((step, index) => {
        const stepNumber = index + 1
        const className = 'step-' + stepNumber + ' background-color-' + Math.floor(30 / steps.length * index)
        const StepComponent = activeStepIndex === index
          ? ActiveStepHeader
          : activeStepIndex > index
            ? ConfirmedStepHeader
            : NotConfirmedStepHeader
        return (
          <StepComponent
            {...step}
            className={className}
            key={step.stepTitle}
            onStepTitleClick={() => setActiveStepIndex(index)}
            stepTitle={stepNumber + ' - ' + step.stepTitle}/>
        )
      })}
    </section>
  )
}

function SelectedLocation({ decimalLatitude, decimalLongitude }) {
  const { t } = useTranslation()

  return (
    <>
      <div>{t('occurrenceForm.locationData.step.selectedData.latitude')}: {decimalLatitude}</div>
      <div>{t('occurrenceForm.locationData.step.selectedData.longitude')}: {decimalLongitude}</div>
    </>
  )
}

SelectedLocation.propTypes = {
  decimalLatitude:  PropTypes.number,
  decimalLongitude: PropTypes.number
}

function renderIdentifiedByLabel({ identifiedBy, institutionCode }) {
  const institutionCodeLabel = institutionCode ? `Institution: ${institutionCode}` : ''
  const identifiedByLabel = identifiedBy.length > 0 ? 'Identified by: ' + identifiedBy.join(', ') : ''
  return [institutionCodeLabel, identifiedByLabel].filter(label => !!label).join('; ')
}

function MeasurementOrFactSummary({ data }) {
  const { t } = useTranslation()

  return <div>{t('occurrenceForm.measurementOrFact.step.title', { number: data.length })}</div>
}

MeasurementOrFactSummary.propTypes = {
  data: PropTypes.array.isRequired
}

function OccurrenceDataSummary({ scientificName, beginDate, endDate }) {
  const occurrenceDataLabel = [
    scientificName,
    format(beginDate, 'D MMMM YYYY'),
    endDate ? ' - ' + format(endDate, 'D MMMM YYYY') : ''
  ].join(' ')

  return (
    <div>{occurrenceDataLabel}</div>
  )
}

OccurrenceDataSummary.propTypes = {
  beginDate:      PropTypes.number,
  endDate:        PropTypes.number,
  scientificName: PropTypes.string
}
