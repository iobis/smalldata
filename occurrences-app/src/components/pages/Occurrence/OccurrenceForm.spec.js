import OccurrenceForm from './OccurrenceForm'
import React from 'react'
import { mount } from 'enzyme/build'

describe('OccurrenceForm', () => {
  it('renders correctly', () => {
    const wrapper = mount(<OccurrenceForm/>)
    expect(wrapper).toMatchSnapshot()
  })
})
