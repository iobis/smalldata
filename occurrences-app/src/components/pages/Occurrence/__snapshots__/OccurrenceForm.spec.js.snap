// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OccurrenceForm renders correctly 1`] = `
<OccurrenceForm>
  <section
    className="section"
  >
    <ActiveStepHeader
      className="step-0"
      dataDescription="occurrenceForm.selectDataset.dataDescription"
      key="occurrenceForm.selectDataset.stepTitle"
      onStepTitleClick={[Function]}
      selectedData="HAB Region 2: Occurrences of harmful (toxic) algal taxa within an area of interest to El Salvador compiled as part of a literature search project."
      stepDescription="occurrenceForm.selectDataset.stepDescription"
      stepTitle="1 - occurrenceForm.selectDataset.stepTitle"
    >
      <StepHeader
        className="active has-background-grey has-text-white"
        dataDescription=""
        iconVisible={false}
        onStepTitleClick={[Function]}
        selectedData=""
        stepDescription="occurrenceForm.selectDataset.stepDescription"
        stepTitle="1 - occurrenceForm.selectDataset.stepTitle"
      >
        <div
          className="step-header columns is-vcentered active has-background-grey has-text-white"
          onClick={[Function]}
        >
          <div
            className="column is-1 is-hidden-mobile"
          >
            <p
              className="is-size-5 is-uppercase"
            >
              <b />
            </p>
          </div>
          <div
            className="column is-7 is-hidden-mobile"
          >
            <p
              className="is-size-5"
            />
          </div>
          <div
            className="column details"
          >
            <div
              className="is-size-6"
            >
              occurrenceForm.selectDataset.stepDescription
            </div>
            <div
              className="is-size-5 is-uppercase"
            >
              <b
                className="step-title"
              >
                1 - occurrenceForm.selectDataset.stepTitle
              </b>
            </div>
          </div>
        </div>
        <SelectDataset
          datasets={
            Array [
              Object {
                "description": "HAB Region 2: Occurrences of harmful (toxic) algal taxa within an area of interest to El Salvador compiled as part of a literature search project.",
                "id": 0,
              },
              Object {
                "description": "NPPSD Short-tailed Albatross Sightings",
                "id": 1,
              },
              Object {
                "description": "PANGAEA - Data from Christian-Albrechts-University Kiel",
                "id": 2,
              },
              Object {
                "description": "NSIS: List of marine benthic algae from Magdalen Islands, Quebec as recorded in 1979",
                "id": 3,
              },
              Object {
                "description": "Seguimiento de 10 crías de tortuga boba nacidas en 2016 en el litoral valenciano, en el marco del Proyecto LIFE 15 IPE ES 012 (aggregated per 1-degree cell)",
                "id": 4,
              },
              Object {
                "description": "Waved Albatross Tracking (aggregated per 1-degree cell)",
                "id": 5,
              },
            ]
          }
          onChange={[Function]}
          selectedDataset={
            Object {
              "description": "HAB Region 2: Occurrences of harmful (toxic) algal taxa within an area of interest to El Salvador compiled as part of a literature search project.",
              "id": 0,
            }
          }
        >
          <div
            className="select-dataset columns is-fluid"
          >
            <table
              className="table is-striped is-fullwidth"
            >
              <tbody>
                <DatasetOption
                  checked={true}
                  description="HAB Region 2: Occurrences of harmful (toxic) algal taxa within an area of interest to El Salvador compiled as part of a literature search project."
                  id={0}
                  key="0"
                  onClick={[Function]}
                >
                  <tr
                    className="dataset-option"
                  >
                    <td>
                      <InputRadio
                        checked={true}
                        name="dataset"
                        onChange={[Function]}
                        value={0}
                      >
                        <input
                          checked={true}
                          className="input-radio is-checkradio"
                          id="dataset-0"
                          name="dataset"
                          onChange={[Function]}
                          type="radio"
                          value={0}
                        />
                        <label
                          htmlFor="dataset-0"
                        />
                      </InputRadio>
                    </td>
                    <td>
                      HAB Region 2: Occurrences of harmful (toxic) algal taxa within an area of interest to El Salvador compiled as part of a literature search project.
                    </td>
                  </tr>
                </DatasetOption>
                <DatasetOption
                  checked={false}
                  description="NPPSD Short-tailed Albatross Sightings"
                  id={1}
                  key="1"
                  onClick={[Function]}
                >
                  <tr
                    className="dataset-option"
                  >
                    <td>
                      <InputRadio
                        checked={false}
                        name="dataset"
                        onChange={[Function]}
                        value={1}
                      >
                        <input
                          checked={false}
                          className="input-radio is-checkradio"
                          id="dataset-1"
                          name="dataset"
                          onChange={[Function]}
                          type="radio"
                          value={1}
                        />
                        <label
                          htmlFor="dataset-1"
                        />
                      </InputRadio>
                    </td>
                    <td>
                      NPPSD Short-tailed Albatross Sightings
                    </td>
                  </tr>
                </DatasetOption>
                <DatasetOption
                  checked={false}
                  description="PANGAEA - Data from Christian-Albrechts-University Kiel"
                  id={2}
                  key="2"
                  onClick={[Function]}
                >
                  <tr
                    className="dataset-option"
                  >
                    <td>
                      <InputRadio
                        checked={false}
                        name="dataset"
                        onChange={[Function]}
                        value={2}
                      >
                        <input
                          checked={false}
                          className="input-radio is-checkradio"
                          id="dataset-2"
                          name="dataset"
                          onChange={[Function]}
                          type="radio"
                          value={2}
                        />
                        <label
                          htmlFor="dataset-2"
                        />
                      </InputRadio>
                    </td>
                    <td>
                      PANGAEA - Data from Christian-Albrechts-University Kiel
                    </td>
                  </tr>
                </DatasetOption>
                <DatasetOption
                  checked={false}
                  description="NSIS: List of marine benthic algae from Magdalen Islands, Quebec as recorded in 1979"
                  id={3}
                  key="3"
                  onClick={[Function]}
                >
                  <tr
                    className="dataset-option"
                  >
                    <td>
                      <InputRadio
                        checked={false}
                        name="dataset"
                        onChange={[Function]}
                        value={3}
                      >
                        <input
                          checked={false}
                          className="input-radio is-checkradio"
                          id="dataset-3"
                          name="dataset"
                          onChange={[Function]}
                          type="radio"
                          value={3}
                        />
                        <label
                          htmlFor="dataset-3"
                        />
                      </InputRadio>
                    </td>
                    <td>
                      NSIS: List of marine benthic algae from Magdalen Islands, Quebec as recorded in 1979
                    </td>
                  </tr>
                </DatasetOption>
                <DatasetOption
                  checked={false}
                  description="Seguimiento de 10 crías de tortuga boba nacidas en 2016 en el litoral valenciano, en el marco del Proyecto LIFE 15 IPE ES 012 (aggregated per 1-degree cell)"
                  id={4}
                  key="4"
                  onClick={[Function]}
                >
                  <tr
                    className="dataset-option"
                  >
                    <td>
                      <InputRadio
                        checked={false}
                        name="dataset"
                        onChange={[Function]}
                        value={4}
                      >
                        <input
                          checked={false}
                          className="input-radio is-checkradio"
                          id="dataset-4"
                          name="dataset"
                          onChange={[Function]}
                          type="radio"
                          value={4}
                        />
                        <label
                          htmlFor="dataset-4"
                        />
                      </InputRadio>
                    </td>
                    <td>
                      Seguimiento de 10 crías de tortuga boba nacidas en 2016 en el litoral valenciano, en el marco del Proyecto LIFE 15 IPE ES 012 (aggregated per 1-degree cell)
                    </td>
                  </tr>
                </DatasetOption>
                <DatasetOption
                  checked={false}
                  description="Waved Albatross Tracking (aggregated per 1-degree cell)"
                  id={5}
                  key="5"
                  onClick={[Function]}
                >
                  <tr
                    className="dataset-option"
                  >
                    <td>
                      <InputRadio
                        checked={false}
                        name="dataset"
                        onChange={[Function]}
                        value={5}
                      >
                        <input
                          checked={false}
                          className="input-radio is-checkradio"
                          id="dataset-5"
                          name="dataset"
                          onChange={[Function]}
                          type="radio"
                          value={5}
                        />
                        <label
                          htmlFor="dataset-5"
                        />
                      </InputRadio>
                    </td>
                    <td>
                      Waved Albatross Tracking (aggregated per 1-degree cell)
                    </td>
                  </tr>
                </DatasetOption>
              </tbody>
            </table>
          </div>
        </SelectDataset>
      </StepHeader>
    </ActiveStepHeader>
    <NotConfirmedStepHeader
      className="step-1"
      dataDescription="Given Values"
      key="Basic Data"
      onStepTitleClick={[Function]}
      selectedData=" 29 April 2019 "
      stepDescription="Mandatory observation information"
      stepTitle="2 - Basic Data"
    >
      <StepHeader
        className="not-confirmed has-text-white step-1"
        dataDescription=""
        iconVisible={false}
        onStepTitleClick={[Function]}
        selectedData=""
        stepDescription="Mandatory observation information"
        stepTitle="2 - Basic Data"
      >
        <div
          className="step-header columns is-vcentered not-confirmed has-text-white step-1"
          onClick={[Function]}
        >
          <div
            className="column is-1 is-hidden-mobile"
          >
            <p
              className="is-size-5 is-uppercase"
            >
              <b />
            </p>
          </div>
          <div
            className="column is-7 is-hidden-mobile"
          >
            <p
              className="is-size-5"
            />
          </div>
          <div
            className="column details"
          >
            <div
              className="is-size-6"
            >
              Mandatory observation information
            </div>
            <div
              className="is-size-5 is-uppercase"
            >
              <b
                className="step-title"
              >
                2 - Basic Data
              </b>
            </div>
          </div>
        </div>
      </StepHeader>
    </NotConfirmedStepHeader>
    <NotConfirmedStepHeader
      className="step-2"
      dataDescription="Main Location"
      key="Location Data"
      onStepTitleClick={[Function]}
      selectedData="North Sea"
      stepDescription="Select the location for data collected"
      stepTitle="3 - Location Data"
    >
      <StepHeader
        className="not-confirmed has-text-white step-2"
        dataDescription=""
        iconVisible={false}
        onStepTitleClick={[Function]}
        selectedData=""
        stepDescription="Select the location for data collected"
        stepTitle="3 - Location Data"
      >
        <div
          className="step-header columns is-vcentered not-confirmed has-text-white step-2"
          onClick={[Function]}
        >
          <div
            className="column is-1 is-hidden-mobile"
          >
            <p
              className="is-size-5 is-uppercase"
            >
              <b />
            </p>
          </div>
          <div
            className="column is-7 is-hidden-mobile"
          >
            <p
              className="is-size-5"
            />
          </div>
          <div
            className="column details"
          >
            <div
              className="is-size-6"
            >
              Select the location for data collected
            </div>
            <div
              className="is-size-5 is-uppercase"
            >
              <b
                className="step-title"
              >
                3 - Location Data
              </b>
            </div>
          </div>
        </div>
      </StepHeader>
    </NotConfirmedStepHeader>
    <NotConfirmedStepHeader
      className="step-3"
      dataDescription="Main Info"
      key="Observation Data"
      onStepTitleClick={[Function]}
      selectedData=""
      stepDescription="Enter further specifics"
      stepTitle="4 - Observation Data"
    >
      <StepHeader
        className="not-confirmed has-text-white step-3"
        dataDescription=""
        iconVisible={false}
        onStepTitleClick={[Function]}
        selectedData=""
        stepDescription="Enter further specifics"
        stepTitle="4 - Observation Data"
      >
        <div
          className="step-header columns is-vcentered not-confirmed has-text-white step-3"
          onClick={[Function]}
        >
          <div
            className="column is-1 is-hidden-mobile"
          >
            <p
              className="is-size-5 is-uppercase"
            >
              <b />
            </p>
          </div>
          <div
            className="column is-7 is-hidden-mobile"
          >
            <p
              className="is-size-5"
            />
          </div>
          <div
            className="column details"
          >
            <div
              className="is-size-6"
            >
              Enter further specifics
            </div>
            <div
              className="is-size-5 is-uppercase"
            >
              <b
                className="step-title"
              >
                4 - Observation Data
              </b>
            </div>
          </div>
        </div>
      </StepHeader>
    </NotConfirmedStepHeader>
    <NotConfirmedStepHeader
      className="step-4"
      dataDescription="DWCA INFO"
      key="Generic Data"
      onStepTitleClick={[Function]}
      selectedData="You have submitted 7 extra fields"
      stepDescription="Enter further specifics"
      stepTitle="5 - Generic Data"
    >
      <StepHeader
        className="not-confirmed has-text-white step-4"
        dataDescription=""
        iconVisible={false}
        onStepTitleClick={[Function]}
        selectedData=""
        stepDescription="Enter further specifics"
        stepTitle="5 - Generic Data"
      >
        <div
          className="step-header columns is-vcentered not-confirmed has-text-white step-4"
          onClick={[Function]}
        >
          <div
            className="column is-1 is-hidden-mobile"
          >
            <p
              className="is-size-5 is-uppercase"
            >
              <b />
            </p>
          </div>
          <div
            className="column is-7 is-hidden-mobile"
          >
            <p
              className="is-size-5"
            />
          </div>
          <div
            className="column details"
          >
            <div
              className="is-size-6"
            >
              Enter further specifics
            </div>
            <div
              className="is-size-5 is-uppercase"
            >
              <b
                className="step-title"
              >
                5 - Generic Data
              </b>
            </div>
          </div>
        </div>
      </StepHeader>
    </NotConfirmedStepHeader>
  </section>
</OccurrenceForm>
`;
